File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int leftHeight = height[0];
        int rightHeight = height[n-1];
        int left = 1;
        int right = n-2;
        int ans =0;
        while(left<=right){
            if(leftHeight<rightHeight){
               leftHeight = Math.max(height[left],leftHeight);
               if(height[left]< leftHeight){
                ans +=leftHeight-height[left];
               }
               left++;
            }
            else{
               rightHeight = Math.max(height[right],rightHeight);
               if(height[right]<rightHeight){
                ans +=rightHeight-height[right];
               }
               right--;
            }
        }
        return ans;
    }
}

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
         int n = height.size();
        int maxLeft = height[0];
        int maxRight = height[n-1];
        int l = 1;
        int r = n-2;
        int ans =0;
        while(l<=r)
        {
            if(maxLeft<maxRight)
            {
                if(height[l]>maxLeft) maxLeft = height[l];
                else ans+= maxLeft-height[l];
                l++;
            }
            else{
                 if(height[r]>maxRight) maxRight = height[r];
                 else ans+= maxRight-height[r];
                 r--;
            }
        }
        return ans;
    }
};